README.md
# express-serve-writable

Express middleware for writing to the filesystem with PUT/DELETE requests.

express-serve-writable is the writable complement to [expressjs/serve-static](https://github.com/expressjs/serve-static)'s static file serving.

This will create a web server that will serve up files in a folder, and let you write or delete files inside that same folder:
```javascript
var express = require("express");
var serveStatic = require("serve-static");
var serveWritable = require("express-serve_writable");

var args = process.argv.slice(2);
var root = args[0];

var app = express();
app.use(serveStatic(root));
app.use(serveWritable(root));
app.listen(3000);
```

## Isn't this a security problem?

Yes.

## What's the use-case for this?

I want to build some local development tools for [my HTML5 video games](http://twoscoopgames.com), and since the games are in HTML, the tools should be too. The problem is that the browser isn't allowed to write to the filesystem, and I need to do that.


lib/CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

## [0.0.1] - 2015-01-04
### Fixed
- Send text/plain Content-Type for 201/403/404 responses.

## [0.0.0] - 2014-12-02
### Added
- Initial release.


lib/serve_writable_test.js
"use strict";

var assert = require("assert");
var FS = require("fs-mock");
var rewire = require("rewire");
var stream = require("stream");

var serveWritable = rewire("./serve_writable");

function mockFs(structure) {
	var fs = new FS(structure);
	serveWritable.__set__("fs", fs);
	return fs;
}

describe("serveWritable", function() {
	describe("with no root", function() {
		it("should explode", function() {
			assert.throws(serveWritable, TypeError);
		});
	});
	describe("with root", function() {
		describe("that doesnt exist", function() {
			it("should explode", function() {
				assert.throws(function() {
					serveWritable("/doesnotexist");
				}, TypeError);
			});
		});
		describe("that does exist", function() {
			it("should return a middleware", function() {
				mockFs({ "dir": {} });
				assert(typeof serveWritable("/dir") === "function");
			});
		});
	});
	describe("middleware", function() {
		var fs, middleware;

		beforeEach(function() {
			fs = mockFs({
				"dir": {
					"todelete": "stuff"
				}
			});
			middleware = serveWritable("/dir");
		});

		describe("with GET request", function() {
			it("should call next", function(done) {
				var req = { method: "GET" };
				var res = {};
				middleware(req, res, done);
			});
		});
		describe("with PUT request", function() {
			it("should write the file", function(done) {
				var req = mockRequest("PUT", "/file", "hello");

				var res = mockResponse();
				res.on("finish", function() {
					assert.equal("hello", fs.readFileSync("/dir/file", { encoding: "utf8" }));
					done();
				});

				middleware(req, res);
			});
			it("should return 201 created", function(done) {
				var req = mockRequest("PUT", "/file", "hello");

				var res = mockResponse();
				res.on("finish", function() {
					assert.equal(201, res.statusCode);
					done();
				});

				middleware(req, res);
			});
			it("should return text/plain mime type", function(done) {
				var req = mockRequest("PUT", "/file", "hello");

				var res = mockResponse();
				res.on("finish", function() {
					assert.equal("text/plain", res.getHeader("Content-Type"));
					done();
				});

				middleware(req, res);
			});
			describe("with path above root", function() {
				it("should return 403 forbidden", function(done) {
					var req = mockRequest("PUT", "/../file", "hello");

					var res = mockResponse();
					res.on("finish", function() {
						assert.equal(403, res.statusCode);
						done();
					});

					middleware(req, res);
				});
				it("should return text/plain mime type", function(done) {
					var req = mockRequest("PUT", "/../file", "hello");

					var res = mockResponse();
					res.on("finish", function() {
						assert.equal("text/plain", res.getHeader("Content-Type"));
						done();
					});

					middleware(req, res);
				});
				it("should not write the file", function(done) {
					var req = mockRequest("PUT", "/../file", "hello");

					var res = mockResponse();
					res.on("finish", function() {
						assert.equal(false, fs.existsSync("/file"));
						done();
					});

					middleware(req, res);
				});
			});
		});
		describe("with DELETE request", function() {
			describe("with existing file", function() {
				it("should delete the file", function(done) {
					var req = mockRequest("DELETE", "/todelete", null);

					var res = mockResponse();
					res.on("finish", function() {
						assert.equal(false, fs.existsSync("/dir/todelete"));
						done();
					});

					middleware(req, res);
				});
				it("should return 204 no content", function(done) {
					var req = mockRequest("DELETE", "/todelete", null);

					var res = mockResponse();
					res.on("finish", function() {
						assert.equal(204, res.statusCode);
						done();
					});

					middleware(req, res);
				});
			});
			describe("with non-existing file", function() {
				it("should return 404 not found", function(done) {
					var req = mockRequest("DELETE", "/doesnotexist", null);

					var res = mockResponse();
					res.on("finish", function() {
						assert.equal(404, res.statusCode);
						done();
					});

					middleware(req, res);
				});
				it("should return text/plain mime type", function(done) {
					var req = mockRequest("DELETE", "/doesnotexist", null);

					var res = mockResponse();
					res.on("finish", function() {
						assert.equal("text/plain", res.getHeader("Content-Type"));
						done();
					});

					middleware(req, res);
				});
			});
		});
	});
});

function mockRequest(method, url, data) {
	var req = stream.Readable();
	req._read = function() {
		req.push(data);
		req.push(null);
	};
	req.method = method;
	req.url = url;
	return req;
}

function mockResponse() {
	var res = stream.Writable();
	res._write = function(chunk, encoding, callback) { // jshint ignore:line
		callback();
	};
	res.headers = {};
	res.setHeader = function(name, value) {
		res.headers[name] = value;
	};
	res.getHeader = function(name) {
		return res.headers[name];
	};
	return res;
}


lib/serve_writable.js
"use strict";

var fs = require("fs");
var path = require("path");

module.exports = function(root) {
	if (!root) {
		throw new TypeError("root is required");
	}
	if (!fs.existsSync(root)) {
		throw new TypeError("root does not exist");
	}

	return function(req, res, next) {
		if (req.method !== "PUT" && req.method !== "DELETE") {
			next();
			return;
		}

		var p = path.normalize(path.join(root, req.url));
		if (p.indexOf(root) !== 0) {
			res.statusCode = 403;
			res.setHeader("Content-Type", "text/plain");
			res.end("403 forbidden");
			return;
		}

		if (req.method === "DELETE") {
			deleteFile(res, p);
			return;
		}

		writeFile(req, res, p);
	};
};

function deleteFile(res, path) {
	fs.unlink(path, function(err) {
		if (err) {
			res.statusCode = 404;
			res.setHeader("Content-Type", "text/plain");
		} else {
			res.statusCode = 204;
		}
		res.end();
	});
}

function writeFile(req, res, path) {
	var createWriteStream = fs.createWriteStream.bind(fs);
	var writable = createWriteStream(path);

	req.on("end", function() {
		res.statusCode = 201;
		res.setHeader("Content-Type", "text/plain");
		res.end("201 created");
	});

	req.pipe(writable);
}
